lib/virtio/virtqueue.c

```diff
@@ -550,8 +566,14 @@ static int vq_ring_enable_interrupt(struct virtqueue *vq, uint16_t ndesc)
 	 * since we last checked. Let our caller know so it processes the new
 	 * entries.
 	 */
-	if (virtqueue_nused(vq) > ndesc) {
-		return 1;
+	if (vq->vq_dev->role == VIRTIO_DEV_MASTER) {
+		if (virtqueue_nused(vq) > ndesc) {
+			return 1;
+		}
+	} else {
+		if (virtqueue_navail(vq) > ndesc) {
+			return 1;
+		}
```

Comment:
It looks like only virtqueue_enable_cb() calls this enable interrupt API,  where  does enable_cb()  get called in this change?

Reply:
Nobody call virtqueue_enable_cb now, since the interrupt is enabled by default. We only need to call virtqueue_disable_cb for the receive queue.
The change just make virtqueue_enable_cb work correctly with both master/slave role, so the result don't surprise us if we call it in the furture.

Reply:
> It looks like only virtqueue_enable_cb() calls this enable interrupt API, where does enable_cb() get called in this change? Should it get called when it wants to know when there is available buffer (that is the buffer is full)?
Master/slave need always enable rx queue callback to response the command quickly, and never need enable tx queue callback, since:
1.each tx queue callback just notify one buffer is returned to the pool
2.if the pool doesn't empty before, we don't need the notification
3.if the pool is empty, rpmsg_virtio_send_offchannel_raw check the buffer available by busy loop
unless we utilize the callback wakeup the send like Linux, we don't need tx queue callback.